vault/src/metrics.rs

```diff
@@ -541,3 +658,476 @@ pub async fn publish_expected_bitcoin_balance<B: BitcoinCoreApi + Clone + Send +
             .set(upperbound as f64 / scaling_factor);
     }
 }
+
+pub async fn publish_tokio_metrics(
+    mut metrics_iterators: HashMap<&str, impl Iterator<Item = TaskMetrics>>,
+) -> Result<(), ServiceError> {
+    let frequency = std::time::Duration::from_millis(TOKIO_POLLING_INTERVAL_MS);
+    loop {
+        for (key, val) in metrics_iterators.iter_mut() {
+            if let Some(task_metrics) = val.next() {
+                let label = HashMap::<&str, &str>::from([(TASK_NAME, *key)]);
+                MEAN_IDLE_DURATION
+                    .with(&label)
+                    .set(task_metrics.mean_idle_duration().as_millis() as i64);
+                MEAN_POLL_DURATION
+                    .with(&label)
+                    .set(task_metrics.mean_poll_duration().as_millis() as i64);
+                MEAN_SCHEDULED_DURATION
+                    .with(&label)
+                    .set(task_metrics.mean_scheduled_duration().as_millis() as i64);
+            }
+        }
+        tokio::time::sleep(frequency).await;
```

Comment:
For the tokio metrics in particular, we need this. The millisecond counters for the various metrics get reset on each read, so the reads must be equally distanced from each other to ensure published data points measure the same thing.

Reply:
I'd still like to see other example usage if possible please - just for my own sanity.

Reply:
This is the example from the `tokio_metrics` crate, the comment mentions that the `intervals()` iterator reads the deltas
https://github.com/tokio-rs/tokio-metrics/blob/main/examples/task.rs#L11

Reply:
@sander2  Check the implementation of intervals here: https://github.com/tokio-rs/tokio-metrics/blob/main/src/task.rs#L1677

Reply:
> @sander2 Check the implementation of `intervals()` here: https://github.com/tokio-rs/tokio-metrics/blob/main/src/task.rs#L1677

Yes I see that the returned values are the _increase since the last poll_. However, we (at least currently) only use the `mean_` functions, which divides by the count. If you poll twice as slow, both the total and the count will be twice as high, but the mean does not change. 

So currently, we _could_ just poll from the prometheus handler directly. However, if the prometheus sampling interval is set too short, the result may be very inconsistent, so that'd be an argument for keeping the sleep approach.

What do you guys think? @gregdhill @savudani8

Reply:
I think the point is to be able to pinpoint when a specific event happened and correlate that with resource usage metrics. The reactive approach would be to read tokio metrics whenever Grafana pings Prometheus, but this means irregularities in the metrics from when Grafana isn't running will be averaged out.